Remove dependency on font awesome and add spinner as SVG

martinsjolanders opened the pull request:
This pull request will remove the dependency on the font awesome spinner by adding a base64 encoded svg that we created with [loading.io ](https://loading.io) and is distributed under [CCO licence](https://en.wikipedia.org/wiki/Creative_Commons_license)
#72

PeterOrneholm commented on src/ActiveLogin.Authentication.BankId.AspNetCore/Areas/BankIdAuthentication/Views/BankId/_LoginScript.cshtml:
```diff
@@ -3,8 +3,9 @@
 <script>
     (function (options) {
         // Pre check
+        console.log("checking status");
```
I guess this is here for debugging :) Please remove.

PeterOrneholm commented on src/ActiveLogin.Authentication.BankId.AspNetCore/Areas/BankIdAuthentication/Views/BankId/_LoginScript.cshtml:
```diff
@@ -3,8 +3,9 @@
         // Pre check
+        console.log("checking status");
         var statusElement = document.querySelector('#bankIdLoginStatus');
-        var statusSpinnerElement = statusElement.querySelector('.status-spinner');
```
I'd still like to keep the class-name status-spinner, because that's still what it is I think :)

PeterOrneholm commented on src/ActiveLogin.Authentication.BankId.AspNetCore/Areas/BankIdAuthentication/Views/BankId/_LoginScript.cshtml:
```diff
@@ -80,8 +79,6 @@
 
         function initialize(requestVerificationToken, returnUrl, loginOptions, personalIdentityNumber, personalIdentityNumberElement) {
             formFieldsetElement.disabled = true;
-            show(formSubmitButtonSpinnerElement);
```
As stated in the general comment, I don't think we should remove the spinner from the button. If we place the image in css (using css background), we can still reuse it.

Placing it in CSS makes it harder to change color though, hmm, if we simplify the generated SVG, extracts the animation into CSS and have the SVG itself inlined we can control the color of the icon using CSS as well.

martinsjolanders commented:
Changed the spinner to be an svg in a partial and added classes to the svg and components in the svg.
This way it is possible to change to spinner totally by adding a new partial that overrides this one or just style the default spinner by setting the styles in the classes status-icon(svg element) or spinner-component (the spinning parts). The user need to nest it with an parent class to only affect the desired spinner, for example 
` .button-status-spinner .status-icon {
            width: 20px;
            height: 20px;
     }`
or 
`.button-status-spinner .status-icon .spinner-component {
            fill: white;
   }`

PeterOrneholm commented on src/ActiveLogin.Authentication.BankId.AspNetCore/Areas/BankIdAuthentication/Views/BankId/_spinner.cshtml:
```diff
@@ -0,0 +1,62 @@
+﻿<svg class="status-icon" width="150px" height="150px" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" viewBox="0 0 100 100" preserveAspectRatio="xMidYMid" style="background: none; shape-rendering: auto;">
```
I ran the SVG [through a minifier](https://jakearchibald.github.io/svgomg/) to make it slightly more compact. Also changed default color to black. Please consider using this one.


```
<svg class="status-icon" width="150" height="150" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 100 100" preserveAspectRatio="xMidYMid" style="background:0 0"><g><rect x="44" y="4" rx="66" ry="6" width="12" height="12" fill="#000000" class="spinner-component"><animate attributeName="opacity" values="1;0" keyTimes="0;1" dur="1s" begin="-0.9166666666666666s" repeatCount="indefinite"/></rect></g><g transform="rotate(30 50 50)"><rect x="44" y="4" rx="66" ry="6" width="12" height="12" fill="#000000" class="spinner-component"><animate attributeName="opacity" values="1;0" keyTimes="0;1" dur="1s" begin="-0.8333333333333334s" repeatCount="indefinite"/></rect></g><g transform="rotate(60 50 50)"><rect x="44" y="4" rx="66" ry="6" width="12" height="12" fill="#000000" class="spinner-component"><animate attributeName="opacity" values="1;0" keyTimes="0;1" dur="1s" begin="-0.75s" repeatCount="indefinite"/></rect></g><g transform="rotate(90 50 50)"><rect x="44" y="4" rx="66" ry="6" width="12" height="12" fill="#000000" class="spinner-component"><animate attributeName="opacity" values="1;0" keyTimes="0;1" dur="1s" begin="-0.6666666666666666s" repeatCount="indefinite"/></rect></g><g transform="rotate(120 50 50)"><rect x="44" y="4" rx="66" ry="6" width="12" height="12" fill="#000000" class="spinner-component"><animate attributeName="opacity" values="1;0" keyTimes="0;1" dur="1s" begin="-0.5833333333333334s" repeatCount="indefinite"/></rect></g><g transform="rotate(150 50 50)"><rect x="44" y="4" rx="66" ry="6" width="12" height="12" fill="#000000" class="spinner-component"><animate attributeName="opacity" values="1;0" keyTimes="0;1" dur="1s" begin="-0.5s" repeatCount="indefinite"/></rect></g><g transform="rotate(180 50 50)"><rect x="44" y="4" rx="66" ry="6" width="12" height="12" fill="#000000" class="spinner-component"><animate attributeName="opacity" values="1;0" keyTimes="0;1" dur="1s" begin="-0.4166666666666667s" repeatCount="indefinite"/></rect></g><g transform="rotate(210 50 50)"><rect x="44" y="4" rx="66" ry="6" width="12" height="12" fill="#000000" class="spinner-component"><animate attributeName="opacity" values="1;0" keyTimes="0;1" dur="1s" begin="-0.3333333333333333s" repeatCount="indefinite"/></rect></g><g transform="rotate(240 50 50)"><rect x="44" y="4" rx="66" ry="6" width="12" height="12" fill="#000000" class="spinner-component"><animate attributeName="opacity" values="1;0" keyTimes="0;1" dur="1s" begin="-0.25s" repeatCount="indefinite"/></rect></g><g transform="rotate(270 50 50)"><rect x="44" y="4" rx="66" ry="6" width="12" height="12" fill="#000000" class="spinner-component"><animate attributeName="opacity" values="1;0" keyTimes="0;1" dur="1s" begin="-0.16666666666666666s" repeatCount="indefinite"/></rect></g><g transform="rotate(300 50 50)"><rect x="44" y="4" rx="66" ry="6" width="12" height="12" fill="#000000" class="spinner-component"><animate attributeName="opacity" values="1;0" keyTimes="0;1" dur="1s" begin="-0.08333333333333333s" repeatCount="indefinite"/></rect></g><g transform="rotate(330 50 50)"><rect x="44" y="4" rx="66" ry="6" width="12" height="12" fill="#000000" class="spinner-component"><animate attributeName="opacity" values="1;0" keyTimes="0;1" dur="1s" begin="0s" repeatCount="indefinite"/></rect></g></svg>
```

martinsjolanders replied:
I changed to use this minified svg instead.

martinsjolanders commented on samples/Standalone.MvcSample/Views/Shared/_Layout.cshtml:
```diff
@@ -61,6 +61,15 @@
             line-height: 60px; /* Vertically center the text there */
         }
+
+        .button-status-spinner .status-icon {
```
reason I name it like that was to enable different styling, but changed to use context styling instead, i.e.  button .status-spinner.

martinsjolanders commented:
Screenshots of the spinners
![button-spinner](https://user-images.githubusercontent.com/13886506/53407103-a9167680-39bb-11e9-919a-a180ff205ae1.jpg)
![spinner](https://user-images.githubusercontent.com/13886506/53407097-a61b8600-39bb-11e9-81b2-c90b907c3800.jpg)

fredrik-lundin commented:
Just to be sure (haven't tested myself), is this tested in all browsers we support? And what browsers Do we support?

I know from experience that svgs, animations, och css in general have a tendency to look and behave differently in different browsers. So just be sure to try it out :)

PeterOrneholm commented:
> Just to be sure (haven't tested myself), is this tested in all browsers we support? And what browsers Do we support?

Great input @fredrik-lundin!

@martinsjolander: Can you try this out in a few browsers? I've made a pen with the spinner here, can you try it out so that it works as expected, that would be awesome: https://codepen.io/PeterOrneholm/pen/QojXzK

Also, I'd like us to give credit to the site that generated the SVG, even if it is a CC0 license. We have a section at the bottom in the Readme.md, would be great to have it there.

martinsjolanders commented:
> > Just to be sure (haven't tested myself), is this tested in all browsers we support? And what browsers Do we support?
> 
> Great input @fredrik-lundin!
> 
> @martinsjolander: Can you try this out in a few browsers? I've made a pen with the spinner here, can you try it out so that it works as expected, that would be awesome: https://codepen.io/PeterOrneholm/pen/QojXzK
> 
After testing this on Browserstack and locally it worked in most browsers except edge and IE 11. I have made some changed to the svg and move the animation to css, and now it works in those browsers as well.
I have testes as followed
**Win10**
- Edge 18 OK
- IE 11 OK
- Firefox 65 OK	
- Chrome 72 OK
- Opera 58 OK

**iPhone 8, iPhone X, iPhone XS**
- Safari OK
- Chrome OK

**Mac OS Yosemite** 
- Firefox 65 OK
- Chrome 72 OK
- Opera 58 OK

**Samsung Galaxy S9**
- Chrome OK
- Firefox OK
- ucbrowser OK

**Samsung Galaxy Note 9**
- Chrome OK
- Firefox OK

**Google Pixel 3**
- Chrome OK
- Firefox OK

**Sony xperia Z5**
- Chrome OK
- Firefox OK

	

> Also, I'd like us to give credit to the site that generated the SVG, even if it is a CC0 license. We have a section at the bottom in the Readme.md, would be great to have it there.

martinsjolanders commented:
> > Just to be sure (haven't tested myself), is this tested in all browsers we support? And what browsers Do we support?
> 
> Great input @fredrik-lundin!
> 
> @martinsjolander: Can you try this out in a few browsers? I've made a pen with the spinner here, can you try it out so that it works as expected, that would be awesome: https://codepen.io/PeterOrneholm/pen/QojXzK
> 
> Also, I'd like us to give credit to the site that generated the SVG, even if it is a CC0 license. We have a section at the bottom in the Readme.md, would be great to have it there.

I have added credits in Readme.md and removed text about font awesome
